ghcide/src/Development/IDE/Core/Shake.hs

```diff
@@ -759,13 +760,14 @@ delayedAction a = do
 -- | Restart the current 'ShakeSession' with the given system actions.
 --   Any actions running in the current session will be aborted,
 --   but actions added via 'shakeEnqueue' will be requeued.
-shakeRestart :: Recorder (WithPriority Log) -> IdeState -> VFSModified -> String -> [DelayedAction ()] -> IO ()
-shakeRestart recorder IdeState{..} vfs reason acts =
+shakeRestart :: Recorder (WithPriority Log) -> IdeState -> VFSModified -> String -> [DelayedAction ()] -> [Key] -> IO ()
+shakeRestart recorder IdeState{..} vfs reason acts keys =
```

Comment:
Perhaps we want something like `IO ()` instead of `Key` so that we can safely perform arbitrary actions in between build sessions, so that things like `SMethod_WorkspaceDidChangeWatchedFiles` which update the state can be safely executed.

Reply:
Or perhaps `IO KeySet`, so that we can perform effects and return a set of keys to be invalidated.

Reply:
make sense, I'll see how is that working out.